include/net/mqtt.h

```diff
@@ -1,474 +1,685 @@
+	enum mqtt_conn_return_code return_code;
+};
+
+/** @brief Parameters for MQTT publish acknowledgment (PUBACK). */
+struct mqtt_puback_param {
+	u16_t message_id;
+};
 
-	/* Internal use only */
-	int (*rcv)(struct mqtt_ctx *ctx, struct net_pkt *);
+/** @brief Parameters for MQTT publish receive (PUBREC). */
+struct mqtt_pubrec_param {
+	u16_t message_id;
+};
+
+/** @brief Parameters for MQTT publish release (PUBREL). */
+struct mqtt_pubrel_param {
+	u16_t message_id;
+};
+
+/** @brief Parameters for MQTT publish complete (PUBCOMP). */
+struct mqtt_pubcomp_param {
+	u16_t message_id;
+};
+
+/** @brief Parameters for MQTT subscription acknowledgment (SUBACK). */
+struct mqtt_suback_param {
+	u16_t message_id;
+	struct mqtt_binstr return_codes;
+};
+
+/** @brief Parameters for MQTT unsubscribe acknowledgment (UNSUBACK). */
+struct mqtt_unsuback_param {
+	u16_t message_id;
+};
+
+/** @brief Parameters for a publish message. */
+struct mqtt_publish_param {
+	/** Messages including topic, QoS and its payload (if any)
+	 *  to be published.
+	 */
+	struct mqtt_publish_message message;
 
-	/** Application type, see: enum mqtt_app */
-	u8_t app_type;
+	/** Message id used for the publish message. Redundant for QoS 0. */
+	u16_t message_id;
 
-	/* Clean session is also part of the MQTT CONNECT msg, however app
-	 * behavior is influenced by this parameter, so we keep a copy here
+	/** Duplicate flag. If 1, it indicates the message is being
+	 *  retransmitted. Has no meaning with QoS 0.
 	 */
-	/** MQTT Clean Session parameter */
-	u8_t clean_session:1;
+	u8_t dup_flag : 1;
 
-	/** 1 if the MQTT application is connected and 0 otherwise */
-	u8_t connected:1;
+	/** Retain flag. If 1, the message shall be stored persistently
+	 *  by the broker.
+	 */
+	u8_t retain_flag : 1;
 };
 
-/**
- * Initializes the MQTT context structure
- *
- * @param ctx MQTT context structure
- * @param app_type See enum mqtt_app
- * @retval 0 always
- */
-int mqtt_init(struct mqtt_ctx *ctx, enum mqtt_app app_type);
+/** @brief List of topics in a subscription request. */
+struct mqtt_subscription_list {
+	/** Array containing topics along with QoS for each. */
+	struct mqtt_topic *list;
 
-/**
- * Release the MQTT context structure
- *
- * @param ctx MQTT context structure
- * @retval 0 on success, and <0 if error
- */
-int mqtt_close(struct mqtt_ctx *ctx);
+	/** Number of topics in the subscription list */
+	u16_t list_count;
+
+	/** Message id used to identify subscription request. */
+	u16_t message_id;
+};
 
 /**
- * Connect to an MQTT broker
- *
- * @param ctx MQTT context structure
- * @retval 0 on success, and <0 if error
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
  */
+union mqtt_evt_param {
+	/** Parameters accompanying MQTT_EVT_CONNACK event. */
+	struct mqtt_connack_param connack;
 
-int mqtt_connect(struct mqtt_ctx *ctx);
+	/** Parameters accompanying MQTT_EVT_PUBLISH event. */
```

Comment:
Should we say here that payload field is to be ignored? Should we define separate struct for received PUBLISH?

Reply:
I'll extend the PUBLISH event description.